op.c

```diff
@@ -16476,24 +16480,81 @@ Perl_rcpv_copy(pTHX_ char *pv) {
 
 /* Subroutine signature parsing */
 
+struct yy_parser_signature_param {
+    UV         argix;        /* positional index of the param */
+    PADOFFSET  padix;        /* pad index of the var holding the param */
+    OP        *defcop;
+    OPCODE     defmode;
+    OP        *defexpr;
+};
+
 struct yy_parser_signature {
-    UV          elems;      /* number of signature elements seen so far */
-    UV          optelems;   /* number of optional signature elems seen */
+    /* A note on terminology. We call each variable that appears in a
+     * signature a "parameter", and each scalar value the caller passes in at
+     * call time we call an "argument".
+     * Each scalar parameter variable will correspond to one argument value.
+     * The complication arises when we consider a slurpy array or hash
+     * parameter. Here, while it is exactly one *parameter*, it may account
+     * zero or more arguments.
+     */
+    UV          next_argix; /* the argument index of the next parameter we add */
+    UV          opt_params; /* number of optional scalar parameters */
     char        slurpy;     /* the sigil of the slurpy var (or null) */
-    OP         *elemops;    /* NULL, or an OP_LINESEQ of individual element and fence ops */
+
+    UV          nparams;     /* number of elements of the following array that are in use */
+    UV          params_size; /* number of elements we could store in the following array */
+    struct yy_parser_signature_param *params;
+
+    OP         *fenceops;     /* NULL, or an OP_LINESEQ containing the ops */
 };
 
 static void
 destroy_subsignature_context(pTHX_ void *p)
 {
     yy_parser_signature *signature = (yy_parser_signature *)p;
 
-    if(signature->elemops)
-        op_free(signature->elemops);
+    if(signature->params) {
+        for(UV parami = 0; parami < signature->nparams; parami++) {
+            struct yy_parser_signature_param *param = signature->params + parami;
+
+            if(param->defcop)
+                op_free(param->defcop);
+            if(param->defexpr)
+                op_free(param->defexpr);
+        }
+        Safefree(signature->params);
+    }
+
+    if(signature->fenceops)
+        op_free(signature->fenceops);
 
     Safefree(signature);
 }
 
+#define subsignature_push_param() S_subsignature_push_param(aTHX)
+static struct yy_parser_signature_param *
+S_subsignature_push_param(pTHX)
+{
+    assert(PL_parser);
+    yy_parser_signature *signature = PL_parser->signature;
+    assert(signature);
+
+    if(signature->nparams >= signature->params_size) {
+        signature->params_size *= 2;
+        Renew(signature->params, signature->params_size, struct yy_parser_signature_param);
```

Comment:
this memblock lasts a very short time on startup right?

So `signature->params_size *= 2;`is over-allocating as a performance optimization right?

I dont see a problem with that, I just want a confirm this over-alloc-ing is intentional and not a underflow/overflow bounds checking accident.